examples/chip-tool/commands/common/CHIPCommand.cpp

```diff
@@ -65,8 +65,20 @@ CHIP_ERROR CHIPCommand::Run()
     ReturnLogErrorOnFailure(mDefaultStorage.Init());
 
     chip::Controller::FactoryInitParams factoryInitParams;
+
     factoryInitParams.fabricIndependentStorage = &mDefaultStorage;
-    uint16_t port                              = mDefaultStorage.GetListenPort();
+
+    // Init group data provider that will be used for all group keys and IPKs for the
+    // chip-tool-configured fabrics. This is OK to do once since the fabric tables
+    // and the DeviceControllerFactory all "share" in the same underlying data.
+    // Different commissioner implementations may want to use alternate implementations
+    // of GroupDataProvider for injection through factoryInitParams.
+    mGroupDataProvider.SetStorageDelegate(&mDefaultStorage);
+    ReturnLogErrorOnFailure(mGroupDataProvider.Init());
+    chip::Credentials::SetGroupDataProvider(&mGroupDataProvider);
+    factoryInitParams.groupDataProvider = &mGroupDataProvider;
```

Comment:
Since the data provider is a singleton accessible with `chip::Credentials::GetGroupDataProvider()`, this parameter seems redundant. Having two ways to access the data provider is a path to inconsistencies. I actually prefer proper injection, keeping the parameter, and removing the global `GetGroupDataProvider()`.

Reply:
Yes, I would like to remove GetGroupDataProvider, but I want to do it as a follow-up, since there is more surgery needed. It will definitely be possible after this. What's missing is having the `SessionManager` get a reference so that it can use the keys properly, without accessing the singleton getter.